lib/sql.js

```diff
@@ -1508,3 +1508,274 @@ SQLConnector.prototype.getInsertedId = function(model, info) {
+};
+
+/*!
+ * Build sql for listing views
+ * @param options {all: for all owners, owner: for a given owner}
+ * @returns {string} The sql statement
+ */
+ // Due to the different implementation structure of information_schema across
+ // connectors, each connector will have to generate its own query
+SQLConnector.prototype.buildQueryViews = function(options) {
+  throw new Error(g.f('{{buildQueryViews}} must be implemented by the connector'));
+};
+
+/**
+ * Discover model definitions
+ *
+ * @param {Object} options Options for discovery
+ * @param {Function} [cb] The callback function
+ */
+SQLConnector.prototype.discoverModelDefinitions = function(options, cb) {
+  if (!cb && typeof options === 'function') {
+    cb = options;
+    options = {};
+  }
+  options = options || {};
+
+  var self = this;
+  var calls = [function(callback) {
+    self.execute(self.buildQueryTables(options), callback);
+  }];
+
+  if (options.views) {
+    calls.push(function(callback) {
+      self.execute(self.buildQueryViews(options), callback);
+    });
+  }
+  async.parallel(calls, function(err, data) {
+    if (err) {
+      cb(err, data);
+    } else {
+      var merged = [];
+      merged = merged.concat(data.shift());
+      if (data.length) {
+        merged = merged.concat(data.shift());
+      }
+      cb(err, merged);
+    }
+  });
+};
+// Due to the different implementation structure of information_schema across
+// connectors, each connector will have to generate its own query
+SQLConnector.prototype.buildQueryColumns = function(schema, table) {
+  throw new Error(g.f('{{buildQueryColumns}} must be implemented by the connector'));
+};
+
+SQLConnector.prototype.buildPropertyType = function(columnDefinition, options) {
+  throw new Error(g.f('{{buildPropertyType}} must be implemented by the connector'));
+};
+
+/*!
+ * Normalize the arguments
+ * @param table string, required
+ * @param options object, optional
+ * @param cb function, optional
+ */
+SQLConnector.prototype.getArgs = function(table, options, cb) {
+  throw new Error(g.f('{{getArgs}} must be implemented by the connector'));
+};
+
+/**
+ * Discover model properties from a table
+ * @param {String} table The table name
+ * @param {Object} options The options for discovery
+ * @param {Function} [cb] The callback function
+ *
+ */
+SQLConnector.prototype.discoverModelProperties = function(table, options, cb) {
+  var self = this;
+  var args = self.getArgs(table, options, cb);
+  var schema = args.schema;
+  var connectorName = self.dataSource.connector.name;
+
+  table = args.table;
+  options = args.options;
+
+  if (typeof(self.getDefaultSchema) === 'function' && !schema) {
+    schema = self.getDefaultSchema();
+  }
+
+
+  // Recommended MySQL 5.7 Boolean scheme. See
+  // http://dev.mysql.com/doc/refman/5.7/en/numeric-type-overview.html
+  // Currently default is the inverse of the recommendation for backward compatibility.
+  var setDefaultOptions = ['mysql'];
```

Comment:
It's fishy to see `mysql` specific code here. If we give each connector a chance to control defaultOptions, we should expose a function such as `prototype.getDefaultOptions()` so that connector implementation can override it.